### ISSUE TITLE
Sorting collections on 1.0.3 leads to data loss

### ISSUE THREAD

**Issue by sharpner** [doctrine/mongodb-odm#1275]
We are currently sorting persistent collections using something like this (pseudo code)
```
$currentPosition = $this->collection->indexOf($item);

$newPosition = $currentPosition + $move;
if ($newPosition < 0) {
    $newPosition = 0;
} elseif ($newPosition >= $this->collection->count()) {
    $newPosition = $this->collection->count() - 1;
}

if ($move < 0) {
    for ($index = $currentPosition; $index > $newPosition; $index--) {
        $this->collection->set($index, $this->collection->get($index - 1));
    }
} else {
    for ($index = $currentPosition; $index < $newPosition; $index++) {
        $this->collection->set($index, $this->collection->get($index + 1));
    }
}

$this->collection->set($newPosition, $item);
```

But since 1.0.3 sorting leads to missing elements between requests. 

I reproduced this writing a test in our software using persist/flush after each sort, which fails on 1.0.3 but succeeds on 1.0.2.

I currently trying to write a test in the mongodb-odm but I need to get into that first...

Any hints/ideas?

**Comment by malarzm** [comment]
Which collection strategy are you using?

**Comment by sharpner** [comment]
```
* @ODM\ReferenceMany(
     *     targetDocument="Item",
     *     cascade={"refresh","persist"},
     *     orphanRemoval="true",
     *     strategy="atomicSet"
     * )
```

**Comment by malarzm** [comment]
Ah crap, no idea then at a glance :/ failing test case will be appreciated

**Comment by sharpner** [comment]
K will give it a try :) thanks for the fast reply.

**Comment by malarzm** [comment]
@sharpner if there is any chance for a failing test case today I'd be able to spend some time tomorrow on fixing it (unless you'd do that for me as well ;) )

**Comment by sharpner** [comment]
I'm on it ;) first time fiddling in the odm, takes me so me time :p

**Comment by sharpner** [comment]
Okay I'm lost. I migrated the test to the odm project, but it succeeds. 

Logically that means the Bug must be in our code, but, a downgrade to 1.0.2 fixes it for us.... Please don't start to work on it, I will have to look into it further on monday. 

Thanks for the help so far.

**Comment by malarzm** [comment]
@sharpner sure, thanks for the heads up!

**Comment by sharpner** [comment]
Okay we've found the culprit. We are using the Gedmo translatable extension. 

After removing the TranslatableListener everything works as expected. It seems the onFlush kills our datasets... Thanks for the support.

**Comment by sharpner** [comment]
Sorry for the confusion, but with refresh the tests will fail.

I looked in the database after executing the tests (with removing tearDown) and validated that those items are in fact missing. Does this help you?

**Comment by malarzm** [comment]
Should do! Will give it a look tomorrow

**Comment by sharpner** [comment]
thx

**Comment by alcaeus** [comment]
@sharpner Thanks for the test cases, I can see what the problem is: orphanRemoval is your enemy on this one, more specifically https://github.com/doctrine/mongodb-odm/commit/02f77ceb2ce577e817f421caea92a989f65e13b6. For 1.0.3 we changed ```set()``` in PersistentCollection to respect orphanRemoval settings. This normally works fine, except that PersistentCollection doesn't care much for duplicate items in a collection, which is what happens in your case. This is why your code worked with 1.0.2 and fails for 1.0.3.

TL;DR: as a workaround, you can replace the items with NULL before switching them:
```
$coll->set(1, null);
$coll->set(2, null);
$coll->set(1, $c);
$coll->set(2, $b);
```


Full explanation of the weird things going on here:

For the following example, let's assume a collection ```[a, b, c]``` and let's assume that we're going to flip elements 1 (```b```) and 2 (```c```).

The first call replaces element 2 with ```b```:
```$coll->set(2, $b)```. This call:
* schedules orphanRemoval for ```c``` since it would be removed from the collection,
* unschedules any pending orphanRemoval for ```b```.
Our collection now looks like this: ```[a, b, b]``` and we have ```c``` scheduled for removal in UnitOfWork.

The second call then replaces element 1 with ```c```:
```$coll->set(1, $c)```. This call:
* schedules orphanRemoval for ```b```,
* unschedules orphanRemoval for ```c```.
Our collection now looks like this: ```[a, c, b]```, which is what we expect. What we don't expect is that pending orphanRemoval for ```b``` in UnitOfWork which will remove the referenced element on the next flush() and cause errors.

**Comment by gruberro** [comment]
I've added another failing test case in https://github.com/doctrine/mongodb-odm/commit/3865bc1b3ac40b10982a484e565fc3a3913205b5. It's based on embedded documents and tries to resort some elements in it's underlaying persistent collection (```EmbedMany```). So it might fail due to some other reasons than stated in https://github.com/doctrine/mongodb-odm/issues/1275#issuecomment-156501140. But I think from a Doctrine-developer perspective it's quite the same issue.

**Comment by alcaeus** [comment]
@gruberro @sharpner I'm closing this to consolidate discussion into #1278, feel free to add any further comments and issues there.
